### PR TITLE
Replace quiver dpi callback with reinit-on-dpi-changed.

### LINKED ISSUE #22804: [Bug]: Quiver not working with subfigure?
### Bug summary

There seems to be an issue when trying to plot quivers in the axes created within subfigures.

### Code for reproduction

```python
import matplotlib.pyplot as plt
fig = plt.figure()
subfigs = fig.subfigures(1, 2)
axsLeft = subfigs[0].subplots(1, 2)
axsLeft[0].quiver(1,1,1,1)
```

### Actual outcome

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/hyzhou/.local/lib/python3.8/site-packages/matplotlib/__init__.py", line 1361, in inner
    return func(ax, *map(sanitize_sequence, args), **kwargs)
  File "/home/hyzhou/.local/lib/python3.8/site-packages/matplotlib/axes/_axes.py", line 5117, in quiver
    q = mquiver.Quiver(self, *args, **kw)
  File "/home/hyzhou/.local/lib/python3.8/site-packages/matplotlib/quiver.py", line 510, in __init__
    self._cid = ax.figure.callbacks.connect('dpi_changed', on_dpi_change)
AttributeError: 'SubFigure' object has no attribute 'callbacks'

### Expected outcome

Expect the same outcome as

```python
fig, ax = plt.subplots(1, 2,)
ax[0].quiver(1,1,1,1)
```

### Additional information

_No response_

### Operating system

Ubuntu

### Matplotlib Version

3.4.3

### Matplotlib Backend

_No response_

### Python version

3.8.5

### Jupyter version

_No response_

### Installation

_No response_

### PR DESCRIPTION
**Pull request by anntzer**
Instead of registering a callback to update itself whenever the dpi
changes, do the same thing as every other artist which depends on dpi
(e.g. FancyArrowPatch), i.e. update oneself as needed.  We can still
save duplicate calculations by only doing the init if the figure dpi
changed compared to last time.

Also drop _new_UV, which is never read.

Closes #22804.

## PR Checklist

**Tests and Styling**
- [ ] Has pytest style unit tests (and `pytest` passes).
- [ ] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).

**Documentation**
- [ ] New features are documented, with examples if plot related.
- [ ] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [ ] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).
- [ ] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).

### CODE REVIEW

**Comment by greglucas** [comment]
Nice! Thanks for the quick fix. I had just started looking into this as well and was going to change to using "resize_event" on the canvas which is what is really desired I think, but this is nicer :) On that note, I assume you realized that "dpi_changed" isn't used anywhere else in the codebase, so I would vote for deprecating that in another PR too... I _think_ `resize_event` should be thrown from the dpi_update function because if you change the dpi you are really resizing the canvas?

**Comment by jklymak** [comment]
Not looking at the details, but should these things be done in trans_dpi_acale?

**Review by QuLogic — APPROVED** [review]

**Comment by QuLogic** [comment]
I'm not sure that that would be possible as these are arrows positioned in data space, but whose size is determined by screen space.

**Comment by jklymak** [comment]
We have this problem relatively often, at least with arrows - we want a glyph to have one dimension in data space, but the shape of the arrow heads in physical space.  This doesn't seem too hard to do in physical space at draw time.  Transfer the anchors to physical space and then draw the arrow and let its transform be physical space (for which we need a better name than trans_dpi_scale) if we do that dpi changes happen at the renderer level and the upper level code doesn't need a callback.

**Comment by jklymak** [comment]
Is this only used for labelsep?  Can we place text with an offset transform?

File: lib/matplotlib/quiver.py [hunk 1]
```diff
@@ -562,18 +525,19 @@ def _init(self):
         """
         # It seems that there are not enough event notifications
         # available to have this work on an as-needed basis at present.
-        if True:  # not self._initialized:
+        if True:  # self._dpi_at_last_init != self.axes.figure.dpi
```
Comment (jklymak) [inline lib/matplotlib/quiver.py:528]: Why did you change the comment?   why does this code get run unconditionally?

File: lib/matplotlib/quiver.py [hunk 2]
```diff
@@ -329,18 +313,16 @@ def on_dpi_change(fig):
...
-        self._cid = None
-        super().remove()  # pass the remove call up the stack
+    @property
+    def labelsep(self):
+        return self._labelsep_inches * self.Q.axes.figure.dpi
 
     def _init(self):
-        if True:  # not self._initialized:
-            if not self.Q._initialized:
+        if True:  # self._dpi_at_last_init != self.axes.figure.dpi
```
Comment (jklymak) [inline lib/matplotlib/quiver.py:324]: I don't understand why this condition is here if its always true?

**Review by jklymak — COMMENTED** [review]
This seems good.  I still think quiver should not have a concept of dpi, but I think this change could go in before a significant refactoring.    Definitely prefer this to a callback.  

have you tested changing the dpi?  I'm not even sure how we do that.

File: lib/matplotlib/quiver.py (hunk 1 above)
Comment (anntzer) [reply lib/matplotlib/quiver.py:528]: I don't know why the original check got commented out (and don't really want to investigate right now), I'm just replacing _initialized by its newer moral equivalent.

File: lib/matplotlib/quiver.py (hunk 2 above)
Comment (anntzer) [reply lib/matplotlib/quiver.py:324]: As above.

**Comment by anntzer** [comment]
Agree that deprecating dpi_changed is perhaps a good idea, but this pr is really just aiming for the minimum fix. Ditto for using offset_transform or dpi_scale_trans instead.
No tests either for now :) (if anyone wants to contribute one...)

**Review by greglucas — APPROVED** [review]
I tested this on my multi-monitor mixed-dpi setup and it looks good to me. I think this is good to go for now and there are a few follow-up things that can be done in subsequent PRs.

**Comment by QuLogic** [comment]
@meeseeksdev backport to v3.5.x

**Comment by QuLogic** [comment]
@meeseeksdev backport to v3.5.x

**Comment by QuLogic** [comment]
meeseeksdev backport to v3.5.x
